doc/dev/osd_internals/op_lock_completion.rst

```diff
@@ -0,0 +1,91 @@
+==================
+OP Lock Completion
+==================
+
+Motivation
+--------
+Current write operation needs acquiring and releasing the PG lock frequently. 
+It causes a huge overhead in the case of heavy random write workload.
+Therefore, this patch focus on minimizing the PG lock contention.
+
+Concepts
+--------
+The main Concept is removing the PG lock when OSD handles completion processing 
+such as committed, applied, receiving reply from replica. OSD will send ack to 
+client with simple counting under the OP lock. Updating metadatas which 
+need the PG are delayed. Delayed works will be re-enqueue to ShardedOpWQ threads or 
+it will be handled existing write operation.
+
+
+*Write Flow*
+A. when a write request is issued (Primary)
+  1. PG lock (acquire) -> 2. do_completion() -> 3. repop queue lock -> 
+  4. add RepGather to repop_queue -> 5. add InProgressOP (ReplicateBackend) or 
+  Op (ECBackend) -> 6. repop queue lock (release) -> 7. queue_transaction() -> 
+  8. PG lock (release)
+
+B. when a write request is issued (Not primary)
+  1. PG lock (acquire) -> 2. do_completion() -> 3. add RepSub to 
+  repop_sub_queue -> 4. add RepModify (ReplicateBackend) or Op (ECBackend) -> 
+  5. queue_transaction() -> 6. PG lock (release)
+
+C. when commit, apply are completed
+  1. complete commit or apply. -> 2. Repop(OP) lock (acquire) -> 
+  3. If all commit or apply are finished, send ack to client -> 
+  4. Repop(OP) lock (release) -> 5. _queue_for_completion()
+
+D. when osd receives MSG_OSD_REPOPREPLY or MSG_OSD_EC_WRITE_REPLY
+  1. repop queue lock -> 2. find out OP -> 3. Repop(OP) lock (acquire) -> 
+  4. check completion (applied or commit) -> 5. If all commit or 
+  apply are finished, send ack to client -> 6. Repop(OP) lock (release) -> 
+  7. _queue_for_completion() -> 8. repop queue lock (release)
+
+E. _queue_for_completion (using ShardedOpWQ threads)
+  1. PG lock (acquire) -> 2. do_completion() (iterate repop_queue and 
+  repop_sub_queue for completion and remove RepGather and RepSub from 
+  repop_queue and repop_sub_queue) -> 3. PG lock (release)
```

Comment:
Reading this it looks like the lock ordering is:

<pre>
- pg lock
  - repop queue lock
    - repop lock
</pre>
The _queue_for_completion description makes it sound like pg lock is taken while repop queue lock is held, it's just doing requeue_front().

Overall this strategy sounds fine, though!  I'm just sorry we took too long and the rebase is going to suck. :(